source/common/api/win32/os_sys_calls_impl.cc

```diff
@@ -0,0 +1,333 @@
+      WSAIoctl(sock, SIO_GET_EXTENSION_FUNCTION_POINTER, &WSARecvMsg_guid, sizeof(WSARecvMsg_guid),
+               &WSARecvMsgFnPtr, sizeof(WSARecvMsgFnPtr), &bytes_received, NULL,
+               NULL) != SOCKET_ERROR,
+      "WSAIoctl SIO_GET_EXTENSION_FUNCTION_POINTER for WSARecvMsg failed, not implemented?");
+
+  closesocket(sock);
+
+  return WSARecvMsgFnPtr;
+}
+
+using WSAMSGPtr = std::unique_ptr<WSAMSG>;
+
+static WSAMSGPtr msghdrToWSAMSG(const msghdr* msg) {
+  WSAMSGPtr wsa_msg(new WSAMSG);
+
+  wsa_msg->name = reinterpret_cast<SOCKADDR*>(msg->msg_name);
+  wsa_msg->namelen = msg->msg_namelen;
+  wsabufResult wsabuf = iovecToWSABUF(msg->msg_iov, msg->msg_iovlen);
+  wsa_msg->lpBuffers = wsabuf.wsabuf_.get();
+  wsa_msg->dwBufferCount = wsabuf.num_vec_;
+  WSABUF control;
+  control.buf = reinterpret_cast<CHAR*>(msg->msg_control);
+  control.len = msg->msg_controllen;
+  wsa_msg->Control = control;
+  wsa_msg->dwFlags = msg->msg_flags;
+
+  return std::move(wsa_msg);
+}
+
+} // namespace
+
+SysCallIntResult OsSysCallsImpl::bind(os_fd_t sockfd, const sockaddr* addr, socklen_t addrlen) {
+  const int rc = ::bind(sockfd, addr, addrlen);
+  return {rc, rc != -1 ? 0 : ::WSAGetLastError()};
+}
+
+SysCallIntResult OsSysCallsImpl::chmod(const std::string& path, mode_t mode) {
+  const int rc = ::_chmod(path.c_str(), mode);
+  return {rc, rc != -1 ? 0 : errno};
+}
+
+SysCallIntResult OsSysCallsImpl::ioctl(os_fd_t sockfd, unsigned long int request, void* argp) {
+  const int rc = ::ioctlsocket(sockfd, request, static_cast<u_long*>(argp));
+  return {rc, rc != -1 ? 0 : ::WSAGetLastError()};
+}
+
+SysCallIntResult OsSysCallsImpl::close(os_fd_t fd) {
+  const int rc = ::closesocket(fd);
+  return {rc, rc != -1 ? 0 : ::WSAGetLastError()};
+}
+
+SysCallSizeResult OsSysCallsImpl::writev(os_fd_t fd, const iovec* iov, int num_iov) {
+  DWORD bytes_sent;
+  wsabufResult wsabuf = iovecToWSABUF(iov, num_iov);
+
+  const int rc =
+      ::WSASend(fd, wsabuf.wsabuf_.get(), wsabuf.num_vec_, &bytes_sent, 0, nullptr, nullptr);
+  if (SOCKET_FAILURE(rc)) {
+    return {-1, ::WSAGetLastError()};
+  }
+  return {bytes_sent, 0};
+}
+
+SysCallSizeResult OsSysCallsImpl::readv(os_fd_t fd, const iovec* iov, int num_iov) {
+  DWORD bytes_received;
+  DWORD flags = 0;
+  wsabufResult wsabuf = iovecToWSABUF(iov, num_iov);
+
+  const int rc = ::WSARecv(fd, wsabuf.wsabuf_.get(), wsabuf.num_vec_, &bytes_received, &flags,
+                           nullptr, nullptr);
+  if (SOCKET_FAILURE(rc)) {
+    return {-1, ::WSAGetLastError()};
+  }
+  return {bytes_received, 0};
+}
+
+SysCallSizeResult OsSysCallsImpl::recv(os_fd_t socket, void* buffer, size_t length, int flags) {
+  const ssize_t rc = ::recv(socket, static_cast<char*>(buffer), length, flags);
+  return {rc, rc != -1 ? 0 : ::WSAGetLastError()};
+}
+
+SysCallSizeResult OsSysCallsImpl::recvmsg(os_fd_t sockfd, msghdr* msg, int flags) {
+  DWORD bytes_received;
+  LPFN_WSARECVMSG WSARecvMsgFnPtr = getFnPtrWSARecvMsg();
```

Comment:
style: variable should be snake_case.

Reply:
Patch incoming, thank you for the review